python/ray/tune/integration/horovod.py

```diff
@@ -0,0 +1,425 @@
+    nics: set = None
+    verbose: int = 1
+    key = secret.make_secret_key() if secret else None
+    ssh_port = None
+    ssh_identity_file = None
+    timeout_s: int = 300
+
+    @property
+    def start_timeout(self):
+        return timeout.Timeout(
+            self.timeout_s,
+            message="Timed out waiting for {activity}. Please "
+            "check connectivity between servers. You "
+            "may need to increase the --start-timeout "
+            "parameter if you have too many servers.")
+
+
+class HorovodMixin:
+    def hostname(self) -> str:
+        # This is probably not the right way to retrieve
+        # the intended hostname.
+        return ray.services.get_node_ip_address()
+
+    def update_env_vars(self, env_vars: Dict[str, str]):
+        """Update the env vars in the actor process."""
+        sanitized = {k: str(v) for k, v in env_vars.items()}
+        os.environ.update(sanitized)
+
+    def env_vars(self):
+        """Check the env vars in the actor process."""
+        return dict(os.environ)
+
+
+@ray.remote
+class NodeColocator:
+    """Responsible for colocation of child actors.
+
+    These actors are given resources equal to the sum of their children.
+    This is a mechanism for gang-scheduling and could be replaced
+    later on with placement groups. Gang-scheduling must occur because
+    otherwise another concurrent group could be placed on this node.
+
+    Right now, the only resources that are explicitly propogated to
+    underlying colocated workers are cuda visible devices.
+    """
+
+    def __init__(self, num_workers: int, use_gpu: bool):
+        self.num_workers = num_workers
+        if use_gpu:
+            gpu_ids = ray.get_gpu_ids(as_str=True)
+            assert len(gpu_ids) == num_workers, gpu_ids
+        self.workers = []
+
+    def create_workers(self, trainable: type, config: dict, logdir: str):
+        """Colocates a number of workers."""
+        # Create a node ip resource label so that we can pin
+        # all of the child actors to the same node. This ensures
+        # colocation and balanced training.
+        node_id = f"node:{ray.services.get_node_ip_address()}"
+        remote_cls = ray.remote(trainable)
+        remote_cls = remote_cls.options(
+            num_cpus=0, num_gpus=0, resources={node_id: 0.01})
+
+        # Create Tune trainable actors.
+        self.workers = [
+            remote_cls.remote(
+                config=config,
+                logger_creator=lambda cfg: logger_creator(cfg, logdir, rank))
+            for rank in range(self.num_workers)
+        ]
+
+        # Propogate cuda visible devices to the underlying
+        # colocated workers.
+        gpu_ids = ray.get_gpu_ids(as_str=True)
+        for worker, gpu_id in zip(self.workers, gpu_ids):
+            worker.update_env_vars.remote({"CUDA_VISIBLE_DEVICES": gpu_id})
+        return node_id
+
+    def get_workers(self) -> List:
+        return self.workers
+
+
+class Coordinator:
```

Comment:
Some of this we can probably move into Horovod so we don't have to expose these internals (which may change) to Ray.

I think it would be good to have a simple `horovod.ray.run` API to use.  Would that be sufficient for Ray Tune?

Reply:
I think so. However, I would need a bit lower level than `spark.run`. Specifically I would need to be able to pass in an arbitrary class and obtain a list of ray Actors with horovod started on them:

```
trainable = wrap_function(self.__class__.function)
assert type(trainable) == type  # Ray Tune specific construct.

# encapsulate logic in horovod repo
actors = hvd.ray.start_actors(trainable, num_workers=100, elastic=False, use_gpu=True)

ray.get([a.method_foo.remote() for a in actors])
```

This is basically what I do in `_HorovodTrainable.setup()`. Note that this actually gives you a lot of flexibility. For example,

```python
class CustomExecutor:
    def execute(self, fn, args):
        return fn(args)

actors = hvd.ray.start_actors(CustomExecutor)

def ray_hvd_run(*args, **kwargs):
    return ray.get([a.execute.remote(*args, **kwargs) for a in actors])

def train_func(args):
    hvd.init()
    ...

# something similar to spark run
ray_hvd_run(train_func, args="foobar")

```

Reply:
This is basically what I do in `_HorovodTrainable.setup()`. If this sounds good, I could easily factor it out.

Reply:
That sounds good.  The two things I would add would be:

1. I think it would be useful to wrap everything in some kind of "Job" object to manage the lifecycle of all the components.
2. We could then build a higher level API on top of this for users who don't need the lower-level control.

So something like:

```
# module horovod.ray

def create_job(num_hosts, num_slots, executor_cls=_default_executor_cls):
    ...

def run(train_fn, args, kwargs, num_hosts, num_slots):
    job = create_job(num_hosts, num_slots)
    try:
        job.start()
        return job.execute(lambda w: w.execute(train_fn, *args, **kwargs))
    finally:
        job.stop()
```

Something like that.  What do you think?  Would that give you enough flexibility for this use case?

Reply:
Yeah, that definitely sounds good; let me push a refactor.